src/betterproto/streams/__init__.py

```diff
@@ -0,0 +1,82 @@
+from .. import Message, encode_varint, decode_varint
+from io import BufferedReader, BufferedWriter
+from typing import Iterable, Optional
+
+
+class CodedOutputStream:
+    stream: BufferedWriter
+
+    def __init__(self, stream: BufferedWriter) -> None:
+        self.stream = stream
+
+    def write_varint(self, value: int) -> None:
+        """Write a single varint value to the stream."""
+        self.stream.write(encode_varint(value))
+
+    def write_message(self, message: Message) -> None:
+        """Write a serialized message to the stream."""
+        self.stream.write(bytes(message))
+
+    def write_messages(self, messages: Iterable[Message]) -> None:
+        """Write multiple messages to the stream."""
+        for message in messages:
+            # Write the size of the message to delimit it and enable parsing
+            self.write_varint(len(bytes(message)))
+            self.write_message(message)
+
+
+class CodedInputStream:
+    stream: BufferedReader
+
+    def __init__(self, stream: BufferedReader) -> None:
+        self.stream = stream
+
+    def read_varint(self) -> Optional[int]:
+        """Read a single varint from the stream. Returns None if stream is at EOF."""
+        byte = self.stream.read(1)
+        if byte == b"":
+            # EOF
+            return None
+
+        # Read extended varint bytes (if varint is longer than 1 byte)
+        value = 0
+        shift = 0
+        while int.from_bytes(byte, byteorder="little") >> 7 == 1:
+            # Add current byte to the value
+            value += (int.from_bytes(byte, byteorder="little") & 0b01111111) << (7 * shift)
+
+            # Read the next byte and check it is valid
+            byte = self.stream.read(1)
+            if byte == b"":
+                # EOF is invalid, as the full varint hasn't been read yet
+                raise ValueError("Stream ended part way through a varint")
+
+            shift += 1
+
+        # Add final byte of the varint to the value
+        value += (int.from_bytes(byte, byteorder="little") & 0b01111111) << (7 * shift)
```

Comment:
This can be simplified to something like
```suggestion
        value = 0
        shift = 0

        while True:
            # Read the next byte and check it is valid
            byte = self.stream.read(1)
            if byte == b"":
                if shift == 0:
                    # EOF is valid
                    return None
                # EOF is invalid, as the full varint hasn't been read yet
                raise ValueError("Stream ended part way through a varint")
                
            # Add current byte to the value
            value += (int.from_bytes(byte, byteorder="little") & 0b01111111) << (7 * shift)
                
            if int.from_bytes(byte, byteorder="little") >> 7 == 0:
                break

            shift += 1
```

Reply:
There is often little reason to do manual count variables in Python.

Use [itertools.count](https://docs.python.org/3/library/itertools.html#itertools.count).

```py
for shift in count():
    ...
```

Try not to do repeated operations, there is no need to do `int.rfrom_bytes` twice, do it once and assign it to a variable:

```py
byte = ...
decoded = int.from_bytes(byte, byteorder="little")
```

Reply:
I'm working on solving this, but this has made me realised that there are no tests for when the stream actually does end too soon, so I'll be adding that alongside this change

Reply:
Updated, let me know if there's any problems with the new changes